Shrink Stokes system assembly

gassmoeller opened the pull request:
This should address #756. It is quite likely that there will be test failures (I only checked incompressible cases without melt so far), but the speedup I have seen has been very nice (~30% faster Stokes matrix assembly, maybe 10-15% for Stokes preconditioner, for a 3d box test model). Given that for simple models the Stokes assembly takes on the order of 10-30% of the total model time this is a nice 3-10% overall speedup for many of my models. In particular simple benchmark problems will benefit.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -929,42 +936,63 @@ namespace aspect
 
+          // First loop over all dofs and find those that are in the Stokes system
+          // save the component (pressure and dim velocities) each belongs to.
+          for (unsigned int k = 0, i = 0; i < stokes_dofs_per_cell; ++k)
```
same as in the other patch discussed today -- I find this difficult to read.

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -1011,46 +1048,46 @@ namespace aspect
+                                                     :
+                                                     eta * 2.0 * (scratch.grads_phi_u[i] * scratch.grads_phi_u[j]))
+                                                    - (use_tensor ?
+                                                       eta_two_thirds * (scratch.div_phi_u[i] * trace(stress_strain_director * scratch.grads_phi_u[j]))
```
hm, does using `eta_two_thirds` really make a difference over `eta * 2 / 3`? There are *so many* multiplications in this expression that it's hard to imagine that introducing an abbreviation makes any kind of difference. This kind of optimization is also something the compiler should extract as a common subexpression hoist.

bangerth replied:
Wow, how completely bizarre. I learn something new every day. Thanks for figuring this out!

With this, this part of the change is obviously completely ok.

bangerth commented:
I think you committed the wrong version that now comments out `eta_two_thirds` and instead uses `*(2./3.)`.

gassmoeller commented:
Looks ok now. I think I addressed all comments, but I would feel better, if somebody could take another look. I had to rewrite the assembler plugins for the free surface and the inner_core test as well (those were the initial test failures).

bangerth commented on source/simulator/assembly.cc:
```diff
@@ -999,6 +1034,11 @@ namespace aspect
+                                             ?
+                                             scratch.material_model_outputs.viscosities[q] * 2.0 / 3.0
+                                             :
+                                             0.0);
```
This value (the 0.0) should not be used, right? If so, can you set it to `std::numeric_limits<double>::quiet_NaN()` like in the statement above? This keeps us on the safe side.

bangerth commented on source/simulator/free_surface.cc:
```diff
@@ -683,57 +689,54 @@ namespace aspect
-              for (unsigned int i=0; i< fe_face_values.dofs_per_cell; ++i)
-                for (unsigned int j=0; j< fe_face_values.dofs_per_cell; ++j)
+              {
+                for (unsigned int i = 0, i_stokes = 0; i_stokes < stokes_dofs_per_cell;)
```
can you add the same comment as elsewhere in the last slot here?

bangerth commented on source/simulator/melt.cc:
```diff
@@ -60,6 +60,29 @@ namespace aspect
+    {
+      template<int dim>
+      bool
+      is_stokes_component (const Introspection<dim> &introspection,
```
I wonder if that's the right name for this function because it returns `true` for both kinds of pressures. I don't recognize this as the "Stokes problem" then. Maybe call it `is_velocity_or_pressures()`?

tjhei replied:
I think that would have been fine, because both pressures are inside the "Stokes part" (it is called assemble_stokes after all).

bangerth commented on source/simulator/melt.cc:
```diff
@@ -230,81 +288,70 @@ namespace aspect
+                                                 :
+                                                 eta * 2.0 * (scratch.grads_phi_u[i] * scratch.grads_phi_u[j]))
+                                                - (use_tensor ?
+                                                   eta * 2.0/3.0 * (scratch.div_phi_u[i] * trace(stress_strain_director * scratch.grads_phi_u[j]))
```
is it worth doing the same here with `eta * 2 / 3`?

bangerth replied:
I guess it'd be ok if you said you don't care too much and leave it to those people who care about melt :)

tjhei commented on source/simulator/free_surface.cc:
```diff
@@ -57,9 +57,20 @@ namespace aspect
+    // and compute our own material_model_inputs in apply_stabilization
+    // (because we want to use the solution instead of the current_linearization_point
+    // to compute the material properties).
+    sim.assemblers->stokes_system_assembler_on_boundary_face_properties.need_face_material_model_data = false;
```
I am not sure this is correct. You are setting it unconditionally to false here, so what if some other code needs the material data?

gassmoeller replied:
You are right, and the line was unnecessary anyway. I removed it. That would have caused trouble in models with free surface and melt (is that even supported? do we have a test for that?).

tjhei replied:
melt + free surface and that would probably not work correctly anyways. But there could be other reasons to assemble boundary terms in the future.

tjhei commented:
I went through the melt and freesurface stuff and it looks good to me. :+1:

gassmoeller commented:
Thanks for the comments. Everything is addressed, I will wait one last time for the tester and then merge.
